src/perf/bin/drvmain.cpp

```diff
@@ -0,0 +1,604 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    QUIC Kernel Mode Performance Driver
+
+--*/
+
+#include <quic_platform.h>
+#include "msquic.h"
+#include <quic_driver_main.h>
+#include "perfioctls.h"
+
+#include "quic_trace.h"
+#ifdef QUIC_CLOG
+#include "driver.cpp.clog.h"
+#endif
+
+#define QUIC_PERF_TAG 'frPQ' // QPrf
+
+DECLARE_CONST_UNICODE_STRING(QuicPerfCtlDeviceName, L"\\Device\\quicperformance");
+DECLARE_CONST_UNICODE_STRING(QuicPerfCtlDeviceSymLink, L"\\DosDevices\\quicperformance");
+
+typedef struct QUIC_DEVICE_EXTENSION {
+    EX_PUSH_LOCK Lock;
+
+    _Guarded_by_(Lock)
+        LIST_ENTRY ClientList;
+    ULONG ClientListSize;
+
+} QUIC_DEVICE_EXTENSION;
+
+WDF_DECLARE_CONTEXT_TYPE_WITH_NAME(QUIC_DEVICE_EXTENSION, QuicPerfCtlGetDeviceContext);
+
+typedef struct QUIC_TEST_CLIENT
+{
+    LIST_ENTRY Link;
+    bool TestFailure;
+
+} QUIC_TEST_CLIENT;
```

Comment:
You should have the stop event in here, to (in the future) allow for multiple user mode clients and each individually be stopped.

Reply:
Yeah I'll add that when I finish up kernel mode.